File: leetcode/101-200/128-longest-consecutive-sequence/bench/longest_consecutive.py
"""Benchmark harness for LeetCode #128 — Longest Consecutive Sequence.

Mirrors longest_consecutive.kara algorithm-for-algorithm. Committed as the
correctness oracle; not a measured lane.
"""

NP = 8
N = 20000
CAPV = 25000
ITERS = 150


def longest_consecutive(nums):
    s = set(nums)
    best = 0
    for v in nums:
        if (v - 1) not in s:
            length = 1
            cur = v
            while (cur + 1) in s:
                cur += 1
                length += 1
            if length > best:
                best = length
    return best


def lcg(seed, n, cap):
    out = []
    x = seed
    for _ in range(n):
        x = (x * 1103515245 + 12345) % 2147483648
        out.append(x % cap)
    return out


def main():
    arrays = [lcg(j + 1, N, CAPV) for j in range(NP)]

    sink = 0
    for it in range(ITERS):
        idx = (it * 3) % NP
        sink += longest_consecutive(arrays[idx])
    print(sink)


if __name__ == "__main__":
    main()

File: leetcode/101-200/128-longest-consecutive-sequence/bench/longest_consecutive.rs
// Benchmark harness for LeetCode #128 — Longest Consecutive Sequence.
// Mirrors longest_consecutive.kara algorithm-for-algorithm.

use std::collections::HashSet;

fn longest_consecutive(nums: &[i64]) -> i64 {
    let mut s: HashSet<i64> = HashSet::new();
    for &v in nums {
        s.insert(v);
    }
    let mut best: i64 = 0;
    for &v in nums {
        if !s.contains(&(v - 1)) {
            let mut length: i64 = 1;
            let mut cur = v;
            while s.contains(&(cur + 1)) {
                cur += 1;
                length += 1;
            }
            if length > best {
                best = length;
            }
        }
    }
    best
}

fn lcg(seed: i64, n: i64, cap: i64) -> Vec<i64> {
    let mut out: Vec<i64> = Vec::new();
    let mut x = seed;
    for _ in 0..n {
        x = (x * 1103515245 + 12345) % 2147483648;
        out.push(x % cap);
    }
    out
}

fn main() {
    let np: i64 = 8;
    let n: i64 = 20000;
    let cap: i64 = 25000;
    let iters: i64 = 150;

    let arrays: Vec<Vec<i64>> = (0..np).map(|j| lcg(j + 1, n, cap)).collect();

    let mut sink: i64 = 0;
    for it in 0..iters {
        let idx = ((it * 3) % np) as usize;
        sink += longest_consecutive(&arrays[idx]);
    }
    println!("{}", sink);
}
